Add no-pull option for image building

anshlykov opened the pull request:
Closes #22679

scottfrederick commented:
@anshlykov Thank you for the contribution. The changes look very good. 

As I was looking at implementing this myself, I started thinking that we could do better than replicating the `--no-pull` option from `pack build`. The semantics of `noPull=false` configuration in Gradle or a `<noPull>false</noPull>` configuration in Maven could be confusing to users because of the double-negative. 

Instead of a boolean "pull or no pull" option, we could model the Kubernetes behavior of a `pullPolicy` option with the possible values `ALWAYS`, `NEVER`, and `IF_NOT_PRESENT` where: 

* `ALWAYS` has the same behavior as your changes with `noPull=false` (the default)
* `NEVER` has the same behavior as your changes with `noPull=true`
* `IF_NOT_PRESENT` checks to see if the image exists in the local daemon and uses it if it does exist, otherwise pulls the image from the remote repository if it does not exist locally

This has the benefits of being familiar to Kubernetes users, providing one additional option for increased control, and making the configuration easier to understand. 

Would you mind updating the PR to implement this option?

anshlykov commented:
@scottfrederick thanks you for the review.

> instead of a boolean "pull or no pull" option, we could model the Kubernetes behavior of a pullPolicy option

It is a good idea. I will implement this option.

anshlykov commented:
@scottfrederick I added the implementation of the `PullPolicy` option. I also refactored the `BuildLog` interface a bit to reduce code duplication. Failed tests in CI are not relevant to this PR as far as I understand

scottfrederick commented on spring-boot-project/spring-boot-tools/spring-boot-buildpack-platform/src/main/java/org/springframework/boot/buildpack/platform/build/AbstractBuildLog.java:
```diff
@@ -41,23 +41,13 @@ public void start(BuildRequest request) {
 	@Override
-	public Consumer<TotalProgressEvent> pullingBuilder(BuildRequest request, ImageReference imageReference) {
-		return getProgressConsumer(" > Pulling builder image '" + imageReference + "'");
+	public Consumer<TotalProgressEvent> pullingImage(ImageReference imageReference, ImageType imageType) {
```
These are public methods of a public class, so we must maintain backward compatibility. Your changes do clean up the interface, and the methods are unlikely to be used outside of the Boot build plugins, but in the interest of compatibility we should leave the existing method signatures in place.

anshlykov replied:
fixed

scottfrederick commented:
@anshlykov Thank you very much for the contribution. This will be a very useful feature. I've merged your changes along with [a commit](https://github.com/spring-projects/spring-boot/commit/6b15822cb13ae9d31bb3b1d0fc8c68c28bb64c71) to polish a few things, update the documentation, and add an integration test.
